File: 4/solution.erl
-module (solution).

main([Arg]) ->
  NumberOfDigit = list_to_integer(Arg),
  Format = "largest palindrome made from the product of two ~p-digit numbers is ~p ~n",
  io:format(Format, [NumberOfDigit, largestPalindromeFromNDigits(NumberOfDigit)]).
  
largestPalindromeFromNDigits(Digit) ->
  MaxNumberOfDigit = trunc(math:pow(10,Digit) - 1),
  Palindromes = [X*Y || X <- lists:seq(1, MaxNumberOfDigit), 
                        Y <- lists:seq(1, MaxNumberOfDigit), 
                        integer_to_list(X*Y) =:= lists:reverse(integer_to_list(X*Y))],
  lists:last(lists:sort(Palindromes)).

File: 4/solution.rb
def palindromic?(product)
  product.to_s == product.to_s.reverse
end

def largest_palindrome(max)
  (0..max).each do |step|
    (0..step).map do |i|
      (max-i) * (max+i-step)
    end.sort.each do |product|
      return product if palindromic?(product)
    end
  end
end

p largest_palindrome(999)